File: 2025/03 March/17 Subset Sum Problem/main.py
# Link: https://www.geeksforgeeks.org/problems/subset-sum-problem-1611555638/1


class Solution:
    def isSubsetSum(self, arr, sum):
        n = len(arr)

        ans = [[False for i in range(sum + 1)] for _ in range(n)]

        for k in range(n):
            ans[k][0] = True
        if arr[0] <= sum:
            ans[0][arr[0]] = True

        for i in range(1, n):
            for j in range(1, sum + 1):
                if j - arr[i] >= 0:
                    ans[i][j] = ans[i - 1][j] or ans[i - 1][j - arr[i]]
                else:
                    ans[i][j] = ans[i - 1][j]
        return ans[n - 1][sum]


if __name__ == "__main__":
    obj = Solution()
    arr = [3, 34, 4, 12, 5, 2]
    sum = 9
    print(obj.isSubsetSum(arr, sum))

    arr = [3, 34, 4, 12, 5, 2]
    sum = 30
    print(obj.isSubsetSum(arr, sum))

    arr = [1, 2, 3]
    sum = 6
    print(obj.isSubsetSum(arr, sum))

File: 2025/03 March/17 Subset Sum Problem/main.cpp
// Link: https://www.geeksforgeeks.org/problems/subset-sum-problem-1611555638/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    bool subsetSumHelper(vector<int> &arr, int sum, int n, vector<vector<int>> &dp)
    {
        if (sum == 0)
            return true;
        if (n == 0)
            return false;

        if (dp[n][sum] != -1)
            return dp[n][sum];

        if (arr[n - 1] > sum)
            return dp[n][sum] = subsetSumHelper(arr, sum, n - 1, dp);

        return dp[n][sum] = subsetSumHelper(arr, sum - arr[n - 1], n - 1, dp) ||
                            subsetSumHelper(arr, sum, n - 1, dp);
    }

    bool isSubsetSum(vector<int> &arr, int sum)
    {
        int n = arr.size();
        vector<vector<int>> dp(n + 1, vector<int>(sum + 1, -1));
        return subsetSumHelper(arr, sum, n, dp);
    }
};

int main()
{
    Solution obj;
    vector<int> arr = {3, 34, 4, 12, 5, 2};
    int sum = 9;
    cout << boolalpha << obj.isSubsetSum(arr, sum) << endl;

    arr = {3, 34, 4, 12, 5, 2};
    sum = 30;
    cout << boolalpha << obj.isSubsetSum(arr, sum) << endl;

    arr = {1, 2, 3};
    sum = 6;
    cout << boolalpha << obj.isSubsetSum(arr, sum) << endl;
    return 0;
}